omegaconf/grammar_visitor.py

```diff
@@ -0,0 +1,357 @@
+                    inputs.append(val)
+                    inputs_str.append(txt)
+            else:
+                assert isinstance(child, TerminalNode)
+
+        assert resolver_name is not None
+        return self.resolver_interpolation_callback(
+            name=resolver_name,
+            inputs=tuple(inputs),
+            inputs_str=tuple(inputs_str),
+        )
+
+    def visitDictKeyValuePair(
+        self, ctx: OmegaConfGrammarParser.DictKeyValuePairContext
+    ) -> Tuple[Any, Any]:
+        from ._utils import _get_value
+
+        assert ctx.getChildCount() == 3  # dictKey COLON element
+        key = self.visit(ctx.getChild(0))
+        colon = ctx.getChild(1)
+        assert (
+            isinstance(colon, TerminalNode)
+            and colon.symbol.type == OmegaConfGrammarLexer.COLON
+        )
+        value = _get_value(self.visitElement(ctx.getChild(2)))
+        return key, value
+
+    def visitListContainer(
+        self, ctx: OmegaConfGrammarParser.ListContainerContext
+    ) -> List[Any]:
+        # BRACKET_OPEN sequence? BRACKET_CLOSE;
+        assert ctx.getChildCount() in (2, 3)
+        if ctx.getChildCount() == 2:
+            return []
+        sequence = ctx.getChild(1)
+        assert isinstance(sequence, OmegaConfGrammarParser.SequenceContext)
+        return list(val for val, _ in self.visitSequence(sequence))  # ignore raw text
+
+    def visitPrimitive(self, ctx: OmegaConfGrammarParser.PrimitiveContext) -> Any:
+        return self._createPrimitive(ctx)
+
+    def visitSequence(
+        self, ctx: OmegaConfGrammarParser.SequenceContext
+    ) -> Generator[Any, None, None]:
+        from ._utils import _get_value
+
+        assert ctx.getChildCount() >= 1  # element (COMMA element)*
+        for i, child in enumerate(ctx.getChildren()):
+            if i % 2 == 0:
+                assert isinstance(child, OmegaConfGrammarParser.ElementContext)
+                # Also preserve the original text representation of `child` so
+                # as to allow backward compatibility with old resolvers (registered
+                # with `legacy_register_resolver()`). Note that we cannot just cast
+                # the value to string later as for instance `null` would become "None".
+                yield _get_value(self.visitElement(child)), child.getText()
+            else:
+                assert (
+                    isinstance(child, TerminalNode)
+                    and child.symbol.type == OmegaConfGrammarLexer.COMMA
+                )
+
+    def visitSingleElement(
+        self, ctx: OmegaConfGrammarParser.SingleElementContext
+    ) -> Any:
+        # element EOF
+        assert ctx.getChildCount() == 2
+        return self.visit(ctx.getChild(0))
+
+    def visitToplevelStr(self, ctx: OmegaConfGrammarParser.ToplevelStrContext) -> str:
+        # (ESC | ESC_INTER | TOP_CHAR | TOP_STR)+
+        return self._unescape(ctx.getChildren())
+
+    def _createPrimitive(
```

Comment:
I just pushed c2624b7 to better match the new Hydra behavior from https://github.com/facebookresearch/hydra/pull/1208, i.e., in particular, interpolations are not allowed anymore as dictionary keys (although I'm still not convinced it was a problem in OmegaConf, it'll be easier to maintain both grammars if they agree on that)

(NB: also pushed 0eef57e which is just moving this method around)

Reply:
Sounds good, thanks.

FYI:
I am still working on the recursive defaults in Hydra. It's a very large pull request and I am delaying reviewing of other large changes in Hydra and OmegaConf until I am done with it.
I will hopefully get it over with early after the holidays.

Thanks for your patience.

Reply:
No worries, thanks for the heads-up!